### PR TITLE
Bigdata cap

### PR DESCRIPTION
**Pull request by albertpastrana**

### CODE REVIEW

File: _posts/2016-05-19-big-data-cap-theorem-plain-english.md [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+
+> Are you a troubled developer because your boss heard about BigData and is asking you about it?
+
+> Are you a CEO who just wants an understandable explanation about the CAP theorem?
+
+> Are you the partner of a developer who just wants to understand what he/she is doing?
+
+Don't worry, we have the solution: BigData and the CAP theorem explained in plain english.
+
+Sometimes we (techies) tend to talk techie and not care about the real world understanding what we are doing. In my particular case, my wife often asks me during dinner about my day. And I have to admit that explaining it in plain (in my case) catalan can be quite challenging.
```
Comment (nathankleyn) [inline _posts/2016-05-19-big-data-cap-theorem-plain-english.md]: I would change the last two sentences a little bit for slightly easier reading:

> In my particular case, my wife often asks me about my day over dinner; I have to admit that explaining it in plain (in my case) catalan can be quite challenging.

File: _posts/2016-05-19-big-data-cap-theorem-plain-english.md [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+
+## Scaling vertically vs horizontally
+
+Imagine that you own a [library](https://en.wikipedia.org/wiki/Library) (yes, that will be our metaphor). You have lots of books and magazines in it but, at some point, you don't have enough space in the shelves, so you decide to buy some more.
+
+<img alt="Adding floors" style="float: right;" src="/assets/images/big-data-cap-theorem/scafolding.jpg">
+
+After some weeks, you've filled the five floors of your library with shelves but you still don't have enough space (disk) for all the books you want to keep. This is when you decide you need to add floors to the library building.
+
+With more books you also need to grow your library index, so you need to buy more cards (memory). You'll probably need to hire more librarians (cpu). Also, there is a point there are queues outside the building (too many visitors) so you need to open more entrance doors (network).
```
Comment (nathankleyn) [inline _posts/2016-05-19-big-data-cap-theorem-plain-english.md]: > there is a point there

should become

> there is a point where there

File: _posts/2016-05-19-big-data-cap-theorem-plain-english.md [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+
+As you have imagined, this is what it's called "scaling up" or "scaling vertically"[^scale]. You find your bottleneck (e.g. memory) and just make it bigger.
```
Comment (nathankleyn) [inline _posts/2016-05-19-big-data-cap-theorem-plain-english.md]: > this is what it's called "scaling up"

should be

> this is what is called "scaling up"

File: _posts/2016-05-19-big-data-cap-theorem-plain-english.md [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+
+Someday you'll see that the cost of expanding your library is prohibitive or, directly that it is physically impossible to add more shelves, indexes or doors. That day is the day you realise you need to build another library.
```
Comment (nathankleyn) [inline _posts/2016-05-19-big-data-cap-theorem-plain-english.md]: I would drop the comma here and the "directly":

> Someday you'll see that the cost of expanding your library is prohibitive or that it is physically impossible to add more shelves, indexes or doors.

File: _posts/2016-05-19-big-data-cap-theorem-plain-english.md [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+
+And yes, your guess is correct, creating a library network and adding more nodes (buildings) to it would be to "scale out" or "scale horizontally" your library.
+
+## Sharding, redundancy and replication factor
+
+### Sharding
+Now that you have multiple buildings you need to take some important decisions. The first one is to decide how to distribute your books among the different libraries (sharding[^shard]).
```
Comment (nathankleyn) [inline _posts/2016-05-19-big-data-cap-theorem-plain-english.md]: I wonder if "take some" should be "make some" here?

File: _posts/2016-05-19-big-data-cap-theorem-plain-english.md [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+
+One possible and easy solution would be to distribute them by author. Books of authors whose surname starts with A, B, C or D go to building number 1. Authors with surname starting with E, F, G or H go to building number 2 and so on.
+
+You soon realise a problem with that approach. The authors surnames are not distributed equally and some authors (like Terry Pratchet) have written many more books than others. The implication is that some of your buildings are half empty while some others are too busy.
+
+You need to choose a better criterion (partition or shard key) that is easy to compute and well distributed. There are several options and there is not a good answer that will work in any scenario (it will depend on how your index is created, what queries can you do, what are the most common queries people is doing...). But, as an example, hashing the book title and using the modulus to decide which building the book is in would provide a better distribution of the books.
```
Comment (nathankleyn) [inline _posts/2016-05-19-big-data-cap-theorem-plain-english.md]: A good "proper" english rule of thumb is to try to never start a sentence with "but". Most of the time, you can swap it for "However":

> However, as an example, hashing the book title and using the modulus to decide which building the book is in would provide a better distribution of the books.

File: _posts/2016-05-19-big-data-cap-theorem-plain-english.md [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+
+One day you read about [what happened to the alexandrians](https://en.wikipedia.org/wiki/Destruction_of_the_Library_of_Alexandria) and think about what would happen if one of the buildings catches fire. So you decide to buy extra copies of your books and put them in different buildings (redundancy).
+
+Of course, this comes with a cost and implications, both of them will depend on how many copies you want to keep (replication factor) among others.
+
+Besides the obvious advantage of having backup copies of each book (you could have achieved that by simply having a warehouse), this new architecture has a major drawback: every time you receive a new edition of a book, you'll need to send the updated copy to each of the buildings with a van, a truck, a train or a bicycle (unfortunately, books don't yet send themselves!)
+
+## CAP Theorem (aka Brewer's theorem)[^cap]
+
+Now that you have several buildings spread all over the city with the books distributed among them is when you'll start having troubles.
```
Comment (nathankleyn) [inline _posts/2016-05-19-big-data-cap-theorem-plain-english.md]: Just trying to make sure this sentence reads correctly, as I think there's something a bit off about the "among them is" part:

> Once you have several buildings spread all over the city with the books distributed among them, you normally find this is when you start having troubles.

File: _posts/2016-05-19-big-data-cap-theorem-plain-english.md [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+
+Let's see what are the problems that arise when you distribute your data by looking at different scenarios but, before that, we will define three desirable conditions that any library should fulfill.
+
+- _Consistent:_ the books (data) in your library should be consistent.
+  That means that if a book has a new edition with some corrections, all the
+  different buildings should have that same edition so that any user requesting
+  the book gets the same information.
+
+- _Available:_ any user requesting a book in your library should be able
+  to get a response regardless it is successful (for example, we have the
```
Comment (nathankleyn) [inline _posts/2016-05-19-big-data-cap-theorem-plain-english.md]: > a response regardless it is successful

should be

> a response regardless of whether it is successful

File: _posts/2016-05-19-big-data-cap-theorem-plain-english.md [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+In the real world, if your system is in two different data-centers you must account for the likelihood of a network partition. The probability of a network problem is much lower if your entire system is in the same datacenter and you control it, but it's not impossible.
+
+The conclusion that comes from all the above is simple. If you want your libraries to be tolerant of not being able to communicate with each other, you'll need to plan what to do in case of a network problem:
+
+- Tell your users the book they are requesting is not available. (Availability)
+
+- Tell your users you have the book they are requesting, but you
+  don't know if it's the latest edition. (Consistency)
+
+This is (again) a bit misleading as your system doesn't need to choose between being consistent or available and the approach taken by most modern solutions is to have a balance between consistency and availability. But that's probably a discussion for another post, this one is already too long.
```
Comment (nathankleyn) [inline _posts/2016-05-19-big-data-cap-theorem-plain-english.md]: Last change, swap "But that's probably" for "However that's probably". 😄

File: _posts/2016-05-19-big-data-cap-theorem-plain-english.md (hunk 1 above)
Comment (albertpastrana) [inline _posts/2016-05-19-big-data-cap-theorem-plain-english.md]: done

File: _posts/2016-05-19-big-data-cap-theorem-plain-english.md (hunk 2 above)
Comment (albertpastrana) [inline _posts/2016-05-19-big-data-cap-theorem-plain-english.md]: done

File: _posts/2016-05-19-big-data-cap-theorem-plain-english.md (hunk 3 above)
Comment (albertpastrana) [inline _posts/2016-05-19-big-data-cap-theorem-plain-english.md]: done

File: _posts/2016-05-19-big-data-cap-theorem-plain-english.md (hunk 4 above)
Comment (albertpastrana) [inline _posts/2016-05-19-big-data-cap-theorem-plain-english.md]: done

File: _posts/2016-05-19-big-data-cap-theorem-plain-english.md (hunk 5 above)
Comment (albertpastrana) [inline _posts/2016-05-19-big-data-cap-theorem-plain-english.md]: done

File: _posts/2016-05-19-big-data-cap-theorem-plain-english.md (hunk 6 above)
Comment (albertpastrana) [inline _posts/2016-05-19-big-data-cap-theorem-plain-english.md]: awesome, thanks :)

File: _posts/2016-05-19-big-data-cap-theorem-plain-english.md (hunk 7 above)
Comment (albertpastrana) [inline _posts/2016-05-19-big-data-cap-theorem-plain-english.md]: 👍
